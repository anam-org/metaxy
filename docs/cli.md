# metaxy

Metaxy - Feature Metadata Management

## Table of Contents

- [`shell`](#metaxy-shell)
- [`migrations`](#metaxy-migrations)
  - [`generate`](#metaxy-migrations-generate)
  - [`scaffold`](#metaxy-migrations-scaffold)
  - [`apply`](#metaxy-migrations-apply)
  - [`status`](#metaxy-migrations-status)
- [`push`](#metaxy-push)

**Usage**:

```console
$ metaxy COMMAND
```

**Arguments**:


**Commands**:

* `migrations`: Metadata migration commands
* `push`: Record all feature versions (push graph snapshot).
* `shell`: Start interactive shell.

## `metaxy shell`

Start interactive shell.

**Usage**:

```console
$ metaxy shell
```

## `metaxy migrations`

Metadata migration commands

**Usage**:

```console
$ metaxy migrations COMMAND
```

**Commands**:

* `apply`: Apply migration(s) up to specified revision.
* `generate`: Generate migration file from detected feature changes.
* `scaffold`: Create an empty migration scaffold for user-defined operations.
* `status`: Show migration status.

## `metaxy migrations generate`

Generate migration file from detected feature changes.

Two modes: 1. Default (no snapshots): Compare store's latest snapshot vs current code 2. Historical (both snapshots): 
Compare two historical snapshots

Automatically detects features that need migration and generates explicit operations for ALL affected features.

Example (default):
    $ metaxy migrations generate  Detected 1 root feature change(s):  ✓ video_processing: abc12345 → def67890  
Generating explicit operations for 3 downstream features:  ✓ feature_c (current: xyz111) ✓ feature_d (current: aaa222) ✓
feature_e (current: bbb333)  Generated 4 total operations (1 root + 3 downstream)
      
Example (historical):
    $ metaxy migrations generate --from-snapshot abc123... --to-snapshot def456...

**Usage**:

```console
$ metaxy migrations generate [OPTIONS]
```

**Options**:

* `--migrations-dir`: Directory for migration files (uses config if not specified)
* `--from-snapshot`: Compare from this historical snapshot ID (optional)
* `--to-snapshot`: Compare to this historical snapshot ID (optional)


## `metaxy migrations scaffold`

Create an empty migration scaffold for user-defined operations.

Generates a migration file template with: - Snapshot IDs from current store state - Empty operations list for manual 
editing - Proper structure and metadata

Use this when you need to write custom migration operations that can't be auto-generated (e.g., complex data 
transformations, backfills).

**Usage**:

```console
$ metaxy migrations scaffold [OPTIONS]
```

**Options**:

* `--migrations-dir`: Directory for migration files (uses config if not specified)
* `--description`: Migration description (optional)
* `--from-snapshot`: Use this as from_snapshot_id (defaults to latest in store)
* `--to-snapshot`: Use this as to_snapshot_id (defaults to current registry)


## `metaxy migrations apply`

Apply migration(s) up to specified revision.

Applies all migrations in dependency order up to the target revision. If no revision specified, applies all migrations.

Migrations are applied with parent validation - parent migrations must be completed before applying child migrations. 
Already-completed migrations are skipped.

Errors if there are multiple heads (migrations with no children) and no revision is specified.

**Usage**:

```console
$ metaxy migrations apply [OPTIONS] [ARGS]
```

**Options**:

* `REVISION, --revision`: Migration ID to apply up to (applies all if not specified)
* `--dry-run, --no-dry-run`: Preview changes without executing  *[default: --no-dry-run]*
* `--force, --no-force`: Re-apply even if already completed  *[default: --no-force]*
* `--migrations-dir`: Directory containing migration files (uses config if not specified)


## `metaxy migrations status`

Show migration status.

Displays all registered migrations and their completion status. Status is derived from system tables (migrations, ops, 
steps).

**Usage**:

```console
$ metaxy migrations status
```


## `metaxy push`

Record all feature versions (push graph snapshot).

Records all features in the active registry to the metadata store with a deterministic snapshot ID. This should be run 
after deploying new feature definitions.

**Usage**:

```console
$ metaxy push [ARGS]
```

**Options**:

### Options

* `STORE, --store`: The metadata store to use. Defaults to the default store.
