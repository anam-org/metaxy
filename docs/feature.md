## Feature Version

Every feature has a **feature version** - hash of its complete specification:
- Feature key
- Field definitions and code versions
- Dependencies (feature-level and field-level)

Being more pedantic, the feature version is calculated by hashing together its field versions and the feature key. The field version is calculated by hashing together the field code version, key, and versions of its dependencies.

Feature and field versions are static and deterministic. They are available as class methods:

```python
class VideoProcessing(Feature, spec=FeatureSpec(
    key=FeatureKey(["video", "processing"]),
    fields=[
        FieldSpec(key=FieldKey(["frames"]), code_version=1),
        FieldSpec(key=FieldKey(["audio"]), code_version=1),
    ],
)):
    pass

# Get feature version
print(VideoProcessing.feature_version())  # "a3f8b2c1"

# Change code_version
class VideoProcessing(Feature, spec=FeatureSpec(
    key=FeatureKey(["video", "processing"]),
    fields=[
        FieldSpec(key=FieldKey(["frames"]), code_version=2),  # Changed!
        FieldSpec(key=FieldKey(["audio"]), code_version=1),
    ],
)):
    pass

print(VideoProcessing.feature_version())  # "d7e9f4a2" (different!)
```

Every metadata row includes a `feature_version` column:
```python
metadata = pl.DataFrame({
    "sample_id": [1, 2, 3],
    "data_version": [...],
    "feature_version": "a3f8b2c1",  # Auto-added by store
})
```

This enables:
- **Current vs historical**: Filter by `feature_version` to get specific versions
- **Migration tracking**: Know which rows need updating
- **Audit trail**: See what feature definition produced each row


## User-Defined Metadata Alignment

When feature definitions change, you may need custom logic to align metadata with upstream changes. Override `Feature.align_metadata_with_upstream`:

### Default Behavior (Inner Join)

```python
class MyFeature(Feature, spec=...):
    # Default behavior: inner join on sample_id
    # Only keeps samples present in ALL upstream features
    pass

# During migration:
# - Reads current metadata
# - Inner joins with upstream on sample_id
# - Only samples present in both current AND upstream survive
```

### Custom Alignment Examples

#### One-to-Many Mapping

```python
class VideoFrames(Feature, spec=FeatureSpec(
    key=FeatureKey(["video", "frames"]),
    deps=[FeatureDep(key=FeatureKey(["videos"]))],
    ...
)):
    @classmethod
    def align_metadata_with_upstream(
        cls,
        current_metadata: pl.DataFrame,
        upstream_metadata: dict[str, pl.DataFrame],
    ) -> pl.DataFrame:
        """Each video produces 30 frames.

        When videos upstream change, regenerate all frame sample IDs.
        """
        video_samples = upstream_metadata["videos"]["sample_id"]

        # Generate frame sample IDs
        frames = []
        for video_id in video_samples:
            for frame_idx in range(30):
                frames.append({
                    "sample_id": f"{video_id}_frame_{frame_idx}",
                    "video_id": video_id,
                    "frame_idx": frame_idx,
                })

        return pl.DataFrame(frames)
```

#### Conditional Filtering

```python
class ProcessedVideos(Feature, spec=...):
    @classmethod
    def align_metadata_with_upstream(
        cls,
        current_metadata: pl.DataFrame,
        upstream_metadata: dict[str, pl.DataFrame],
    ) -> pl.DataFrame:
        """Only process videos longer than 10 seconds."""

        videos = upstream_metadata["videos"]

        # Filter by duration
        valid_videos = videos.filter(pl.col("duration") > 10)

        # Keep custom columns from current_metadata if available
        if len(current_metadata) > 0:
            return current_metadata.join(
                valid_videos.select(pl.col("sample_id")),
                on="sample_id",
                how="inner",  # Only keep samples that pass filter
            )
        else:
            return valid_videos.select(["sample_id", "duration", "path"])
```

#### Outer Join (Keep All Samples)

```python
class MergedFeature(Feature, spec=...):
    @classmethod
    def align_metadata_with_upstream(
        cls,
        current_metadata: pl.DataFrame,
        upstream_metadata: dict[str, pl.DataFrame],
    ) -> pl.DataFrame:
        """Keep union of all upstream samples (outer join)."""

        all_samples = set()
        for upstream_df in upstream_metadata.values():
            all_samples.update(upstream_df["sample_id"].to_list())

        return pl.DataFrame({"sample_id": sorted(all_samples)})
```
