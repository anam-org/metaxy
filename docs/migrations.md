# Metadata Migration System

## Overview

Metaxy's migration system enables safe, automated updates to feature metadata when feature definitions change. Migrations are explicit, idempotent, and automatically propagate through the entire dependency graph by default. They can be customized and user-defined. Migrations respect custom `load_input` method which may be implemented for specific features.

**Prerequisites:** The migration system requires that you record feature graph snapshots in your CD (Continuous Deployment) workflow using `metaxy push`. This command snapshots the complete feature graph state and is essential for migration detection to work.

## Core Concepts

### System Tables

Metaxy maintains two system tables in the metadata store:

#### 1. Feature Version History
**Key:** `["metaxy-system", "feature_versions"]`

Tracks when each feature version was recorded:

```python
{
    "feature_key": "video_processing",
    "feature_version": "a3f8b2c1",
    "recorded_at": datetime(2025, 1, 10, 14, 30, 0),
    "fields": ["frames", "audio"],
    "snapshot_version": "9fee05a3",  # Graph snapshot hash (if recorded with record_feature_graph_snapshot)
}
```

**Purpose:**
- Detection of latest version for migrations
- Audit trail of materializations
- Enables `detect_feature_changes()` to work

**Updated explicitly** by calling `store.record_feature_graph_snapshot(Feature)`.
This is typically done in CI after successful materialization.

#### 2. Migration History
**Key:** `["metaxy-system", "migrations"]`

Tracks applied migrations:

```python
{
    "migration_id": "migration_20250113_103000",
    "applied_at": datetime(2025, 1, 13, 10, 30, 0),
    "status": "completed",  # or "partial", "failed"
    "operations_count": 2,
    "affected_features": ["video_processing", "feature_c", "feature_d"],
    "errors": None,  # or dict of errors if partial/failed
}
```

**Purpose:**
- Idempotency - prevent re-running completed migrations
- Recovery tracking - know which migrations are partial
- Audit trail

## Recording Feature Graph Snapshots (User Responsibility)

**YOU MUST** record feature graph snapshots in your CD workflow for migrations to work.

### Using `metaxy push` in CD

Add this to your deployment pipeline (BEFORE materializing features):

```bash
# In your CI/CD pipeline (GitHub Actions, Jenkins, etc.)
$ metaxy push
```

This command:
1. Serializes the complete feature graph (all features, dependencies, field definitions)
2. Generates a deterministic `snapshot_version` (hash of all feature_versions)
3. Records the snapshot in the `__metaxy__/feature_versions` system table
4. Returns the snapshot_version for your records

**Example CD Workflow:**
```bash
#!/bin/bash
# deploy.sh - Your deployment script

# 1. Deploy new code to production
git pull origin main
pip install -e .

# 2. Record feature graph snapshot (CRITICAL!)
metaxy push
# Output: "Recorded snapshot: a3f8b2c1... (12 features)"

# 3. Run your feature materialization pipeline
python -m myproject.compute_features

# 4. Generate migration if code changed
metaxy migrations generate
# If features changed, creates: migrations/migration_20250113_103000.yaml

# 5. Apply migrations automatically (discovers all pending migrations)
metaxy migrations apply
```

**Why is `metaxy push` required?**
- Migration detection compares "latest snapshot in store" vs "current code"
- Without snapshots, the system cannot detect what changed
- Provides audit trail: what feature versions were deployed when
- Enables historical graph reconstruction for safe migrations

**What gets recorded:**
```python
# For each feature in the graph:
{
    "feature_key": "video_processing",
    "feature_version": "a3f8b2c1",  # Hash of feature definition
    "snapshot_version": "9fee05a3",      # Hash of entire graph
    "recorded_at": "2025-01-13T10:30:00Z",
    "feature_spec": {...},          # Complete feature definition (serialized)
    "feature_class_path": "myproject.features.VideoProcessing"
}
```

**Snapshot version Properties:**

The `snapshot_version` is a deterministic hash representing the entire feature graph state:
- ✅ Idempotent - same graph = same snapshot_version every time
- ✅ Unique - any change = different snapshot_version
- ✅ No timestamp - purely content-based hash
- ✅ Groups all features deployed together

**Benefits:**
- Detect exactly which features changed between deployments
- Reconstruct historical feature graphs for safe migrations
- Audit trail of what was deployed when
- Identify incomplete deployments (some features missing snapshot)

## Migration Workflow (Complete Example)

### Step 1: Initial State (Production)

Your CD pipeline has already run `metaxy push` to record the current feature graph:

```bash
# In production CD (previously run)
$ metaxy push
# Recorded snapshot: a3f8b2c1... (15 features)
```

### Step 2: Developer Refactors Feature Code (No Computation Change)

```python
# Before: STT feature depends on entire video (inefficient graph)
class SpeechTranscription(Feature, spec=FeatureSpec(
    key=FeatureKey(["speech", "transcription"]),
    deps=[
        FeatureDep(key=FeatureKey(["video", "file"]))  # Entire video!
    ],
    fields=[
        FieldSpec(
            key=FieldKey(["text"]),
            code_version=1,
            deps=[FieldDep(
                feature_key=FeatureKey(["video", "file"]),
                fields=[FieldKey(["default"])]  # Uses whole video
            )]
        )
    ]
)):
    pass

# After: STT only depends on audio field (more precise, same computation!)
class SpeechTranscription(Feature, spec=FeatureSpec(
    key=FeatureKey(["speech", "transcription"]),
    deps=[
        FeatureDep(key=FeatureKey(["video", "file"]), fields=["audio"])  # Only audio!
    ],
    fields=[
        FieldSpec(
            key=FieldKey(["text"]),
            code_version=1,  # UNCHANGED - computation is same
            deps=[FieldDep(
                feature_key=FeatureKey(["video", "file"]),
                fields=[FieldKey(["audio"])]  # More precise dependency
            )]
        )
    ]
)):
    pass

# This changes feature_version (because deps changed) but NOT the actual transcription
# Perfect use case for DataVersionReconciliation!
```

### Step 3: Generate Migration

```bash
# Generate migration by comparing latest snapshot (a3f8b2c1...) vs current code
$ metaxy migrations generate

# Output:
# From: latest snapshot a3f8b2c1...
# To: current active graph (snapshot def67890...)
#
# Detected 1 root feature change(s):
#   ✓ speech/transcription
#
# Generating explicit operations for 2 downstream feature(s):
#   ✓ speech/diarization
#   ✓ speech/analysis
#
# Generated 3 total operations (1 root + 2 downstream)
# Created: migrations/migration_20250113_103000.yaml
```

The generated migration file contains explicit operations for ALL affected features.

### Step 4: Review Migration YAML

```bash
$ cat migrations/migration_20250113_103000.yaml
```

See the [Migration File Format](#migration-file-format-yaml) section below for the complete structure.

### Step 5: Deploy and Apply

```bash
# In your CD pipeline
$ metaxy push  # Record new snapshot (def67890...)
$ metaxy migrations apply  # Auto-discovers migration files

# Output:
# Discovered 1 migration file(s)
# Applying migration_20250113_103000...
#   ✓ reconcile_stt_transcription: 1,234 rows affected
#   ✓ reconcile_speaker_diarization: 856 rows affected
#   ✓ reconcile_speech_analysis: 2,100 rows affected
# Migration completed in 2.3s
```

## Migration Execution Details

### How Migrations Work (Immutable Copy-on-Write)

**Migrations preserve immutability** - they never modify existing metadata.
Instead, they **copy** old metadata, recalculate data versions, and **append** new rows.

When you run `metaxy migrations apply`:

**Step 1: Check Idempotency**
```python
# Is migration already completed?
if _is_migration_completed(store, migration.id):
    print("Migration already completed, skipping.")
    return
```

**Step 2: Migrate Source Features (Immutable Copy-on-Write)**
```python
for operation in migration.operations:
    # Operations derive old/new feature_versions from migration's snapshot versions
    # Query feature versions from snapshot metadata
    from_feature_version = get_version_from_snapshot(
        store, operation.feature_key, migration.from_snapshot_version
    )
    to_feature_version = get_version_from_snapshot(
        store, operation.feature_key, migration.to_snapshot_version
    )

    # Query rows with OLD feature_version
    old_rows = store.read_metadata(
        feature,
        current_only=False,
        filters=pl.col("feature_version") == from_feature_version,
    )

    if len(old_rows) == 0:
        # Already migrated (idempotent)
        continue

    # Copy ALL metadata columns (preserves paths, labels, etc.)
    # Exclude only data_version, feature_version, and snapshot_version (will be recalculated)
    columns_to_keep = [c for c in old_rows.columns
                      if c not in ["data_version", "feature_version", "snapshot_version"]]
    sample_metadata = old_rows.select(columns_to_keep)

    # Recalculate data versions and APPEND new rows
    # Old rows remain unchanged (immutable!)
    # New rows get:
    # - Same sample_uid and user metadata columns
    # - Recalculated data_version (based on new feature definition)
    # - New feature_version and snapshot_version
    store.calculate_and_write_data_versions(
        feature=feature,
        sample_df=sample_metadata,
    )
```

**Result:** Metadata store now contains BOTH old and new versions:
- Old rows: `feature_version=abc123`, `snapshot_version=snap1`, original `data_version`, original metadata
- New rows: `feature_version=def456`, `snapshot_version=snap2`, recalculated `data_version`, **same metadata**

Reading with `current_only=True` returns only new rows.

**Step 3: Process All Operations (Explicit in YAML)**
```python
# All operations are explicit in the migration YAML
# Both root changes AND downstream reconciliations are listed
# No implicit propagation - everything is explicit and trackable

for operation in migration.operations:
    # Each operation processes one feature
    # Example: "reconcile_speaker_diarization" depends on "reconcile_stt_transcription"

    if isinstance(operation, DataVersionReconciliation):
        # For features with upstream dependencies:
        # 1. Load existing metadata (old feature_version)
        # 2. Recalculate data_versions based on upstream changes
        # 3. Write new rows with updated feature_version + snapshot_version
        operation.execute(
            store,
            from_snapshot_version=migration.from_snapshot_version,
            to_snapshot_version=migration.to_snapshot_version
        )
```

**Key Insight:** All affected features (root + downstream) are explicitly listed in the YAML.
The generator creates operations for ALL features that need reconciliation, making the
migration fully transparent and auditable.

**Step 4: Record Completion**
```python
# Write to migration history
migration_record = pl.DataFrame({
    "migration_id": migration.id,
    "applied_at": datetime.now(),
    "status": "completed",
    "operations_count": len(migration.operations),
    "affected_features": [...],
})

store.write_metadata(
    FeatureKey(["metaxy-system", "migrations"]),
    migration_record,
)
```

### Failure Recovery

Migrations are **automatically recoverable** from partial failures.

**How recovery works:**
1. Re-run queries for rows with old `feature_version`
2. If not found: already migrated, skip
3. If found: migrate them
4. Eventually all rows migrated, migration marked completed
5. Future runs are instant no-ops

Migrations can be safely re-run multiple times.

## Migration File Format (YAML)

```yaml
version: 1  # Migration schema version
id: "migration_20250113_103000"  # Unique ID (timestamp-based)
parent_migration_id: "migration_20250110_120000"  # Previous migration (if any)
description: "Auto-generated migration for 2 changed features + 3 downstream"
created_at: "2025-01-13T10:30:00Z"

# Snapshot versions identify the complete feature graph state (before and after)
from_snapshot_version: "a3f8b2c1..."  # Source snapshot (old state in store)
to_snapshot_version: "def67890..."    # Target snapshot (new state in code)

operations:
  # Root features that changed (code refactoring, not computation changes)
  - id: "reconcile_stt_transcription"
    type: "metaxy.migrations.ops.DataVersionReconciliation"
    feature_key: ["speech", "transcription"]
    reason: "Fixed dependency: now depends only on audio field instead of entire video. Computation logic unchanged, just cleaner graph structure."

  - id: "reconcile_face_detection"
    type: "metaxy.migrations.ops.DataVersionReconciliation"
    feature_key: ["video", "face_detection"]
    reason: "Refactored field structure from single 'default' to 'faces' and 'confidence' fields. Same underlying model, just better schema."

  # Downstream features (auto-generated)
  - id: "reconcile_speaker_diarization"
    type: "metaxy.migrations.ops.DataVersionReconciliation"
    feature_key: ["speech", "diarization"]
    reason: "Reconcile data_versions due to changes in: speech/transcription"

# Feature versions are derived from snapshot versions at runtime
# Each snapshot uniquely identifies all feature versions in the graph

# IMPORTANT: DataVersionReconciliation is for when CODE changed but COMPUTATION didn't
# Use cases:
# - Dependency graph refactoring (more precise dependencies, merging features)
# - Field structure changes (renaming, splitting fields)
#
# If computation ACTUALLY changed, users must re-run their pipeline
# Migrations cannot recreate lost computation results
```

### Migration File Location

Migrations are stored in a directory (default: `migrations/`) and named by timestamp:

```
migrations/
├── 20250110_120000_initial_setup.yaml
├── 20250113_103000_update_video_processing_audio_processing.yaml
└── 20250115_160000_add_new_features.yaml
```

Files are applied **in sorted order** (timestamp ensures correct ordering).

## Best Practices

### 1. Always Dry-Run First

```bash
$ metaxy migrations apply --dry-run
# Review impact before executing
$ metaxy migrations apply
```

### 2. Commit Migrations to Git

Migrations are part of your codebase:
```bash
$ git add migrations/
$ git commit -m "Add migration for video processing v2"
```

### 3. Understand When to Use DataVersionReconciliation

**DataVersionReconciliation is ONLY for when:**
- ✅ Code structure changed (refactoring, better dependencies)
- ✅ Graph topology changed (more precise field dependencies)
- ✅ Schema improved (field renaming, splitting)
- ✅ **Computation logic is unchanged** - results would be identical

**Do NOT use DataVersionReconciliation when:**
- ❌ Algorithm actually changed (different output)
- ❌ Model updated (new ML model version)
- ❌ Bug fixed that affects results
- ❌ **Computation would produce different results**

**Example - Use DataVersionReconciliation:**
```yaml
# Before: STT depends on entire video
deps: [FeatureDep(key=["video", "file"])]

# After: STT depends only on audio (more precise)
deps: [FeatureDep(key=["video", "file"], fields=["audio"])]

# Computation is identical, just cleaner graph
reason: "Refined dependency to audio field only. Transcription logic unchanged."
```

**Example - Re-run Pipeline Instead:**
```python
# Before: Whisper v2
model = whisper.load_model("base")

# After: Whisper v3 (DIFFERENT RESULTS!)
model = whisper.load_model("large-v3")

# DO NOT USE MIGRATION - re-run your pipeline!
# Results would be different, existing data is now stale
```

### 4. Document Reasons

Always fill in meaningful `reason` fields explaining what changed:
```yaml
reason: "Corrected dependency from entire video to audio field only. Computation unchanged."  # ✓ Good
reason: "Refactored field structure - split 'default' into 'faces' and 'confidence'. Same detection model."  # ✓ Good
reason: "TODO: Describe what changed"  # ✗ Bad (leftover template)
reason: "Updated algorithm"  # ✗ Bad (vague - did results change?)
```

### 5. Test Alignment Logic

If you override `load_input()`, test it thoroughly:
```python
def test_custom_alignment():
    current = pl.DataFrame({"sample_uid": [1, 2, 3], "custom_field": ["a", "b", "c"]})
    upstream = {"videos": pl.DataFrame({"sample_uid": [2, 3, 4]})}

    result = MyFeature.load_input(current, upstream)

    # Verify alignment behavior
    assert set(result["sample_uid"].to_list()) == {2, 3}  # Inner join
    assert "custom_field" in result.columns  # Preserved
```

### 6. Monitor Migration Progress

For large migrations, monitor progress:
```python
result = apply_migration(store, migration)
print(f"Affected {len(result.affected_features)} features")
print(f"Duration: {result.duration_seconds}s")
if result.errors:
    print(f"Errors: {result.errors}")
```

### 7. Keep Old Versions

Don't manually delete old feature_version rows unless you have a specific cleanup policy. They serve as:
- Historical record
- Debug reference
- Rollback option
