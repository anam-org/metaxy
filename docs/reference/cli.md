# CLI Commands

This section provides a comprehensive reference for all Metaxy CLI commands.

::: cyclopts
    :module: metaxy.cli.app:app
    :heading-level: 2
    :recursive: true
    :flatten-commands: false
    :generate-toc: true

## Examples

### Recording a graph snapshot

```bash
# Push the current feature graph to the metadata store
metaxy graph push
```

The recommendation is to run this command in your CD pipeline.

### Generating and applying migrations

```bash
# Generate a migration for detected changes
metaxy migrations generate --op metaxy.migrations.ops.DataVersionReconciliation

# Apply pending migrations
metaxy migrations apply
```

### Visualizing the feature graph

```bash
# Render as terminal tree view
metaxy graph render

# Render as Mermaid diagram
metaxy graph render --format mermaid

# Compare two snapshots
metaxy graph-diff render <snapshot-id> current --format mermaid
```
