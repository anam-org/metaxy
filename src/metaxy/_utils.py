import narwhals as nw
import polars as pl


def collect_to_polars(lazy_frame: nw.LazyFrame) -> pl.DataFrame:
    """Helper to collect a Narwhals LazyFrame and convert to Polars DataFrame.

    This handles all backend conversions (Polars, DuckDB/PyArrow, etc.) transparently.
    Used throughout tests for materializing query results.
    """
    return lazy_frame.collect().to_polars()
