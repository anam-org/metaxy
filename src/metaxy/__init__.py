from metaxy.models.container import ContainerDep, ContainerSpec, SpecialContainerDep
from metaxy.models.feature import Feature
from metaxy.models.feature_spec import FeatureDep, FeatureDepMetadata, FeatureSpec
from metaxy.models.types import ContainerKey, FeatureKey

__all__ = [
    "Feature",
    "FeatureDep",
    "FeatureDepMetadata",
    "FeatureSpec",
    "ContainerDep",
    "ContainerSpec",
    "SpecialContainerDep",
    "FeatureKey",
    "ContainerKey",
]
