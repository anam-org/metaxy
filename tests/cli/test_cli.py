"""Tests for CLI commands."""

from datetime import datetime
from pathlib import Path

import polars as pl
import pytest
from syrupy.assertion import SnapshotAssertion

from metaxy import (
    FeatureKey,
    FeatureSpec,
    FieldKey,
    FieldSpec,
)
from metaxy._testing import TempFeatureModule
from metaxy._utils import collect_to_polars
from metaxy.cli.app import app
from metaxy.cli.context import set_config
from metaxy.cli.migrations import app as migrations_app
from metaxy.models.feature import FeatureGraph


@pytest.fixture
def cli_graph():
    """Test graph with features using temp module for importability."""
    temp_module = TempFeatureModule("test_cli_features")

    test_feature_spec = FeatureSpec(
        key=FeatureKey(["test_cli", "feature"]),
        deps=None,
        fields=[FieldSpec(key=FieldKey(["default"]), code_version=1)],
    )

    temp_module.write_features({"TestFeature": test_feature_spec})
    graph = temp_module.get_graph()

    yield graph

    temp_module.cleanup()


@pytest.fixture
def cli_config_file(tmp_path: Path, cli_graph: FeatureGraph) -> Path:
    """Create test config file."""
    config_file = tmp_path / "metaxy.toml"
    config_file.write_text("""
store = "test"

[stores.test]
type = "metaxy.metadata_store.InMemoryMetadataStore"

[stores.test.config]
""")
    return config_file


def test_cli_help():
    """Test main CLI help."""
    # Use parse_args to avoid sys.exit
    app.parse_args(["--help"], print_error=False, exit_on_error=False)
    # Should not raise


def test_migrations_help():
    """Test migrations subcommand help."""
    migrations_app.parse_args(["--help"], print_error=False, exit_on_error=False)
    # Should not raise


def test_migrations_generate_help():
    """Test migrations generate help."""
    migrations_app.parse_args(
        ["generate", "--help"], print_error=False, exit_on_error=False
    )
    # Should not raise


def test_migrations_generate_no_changes(
    tmp_path: Path,
    cli_graph: FeatureGraph,
    capsys,
    snapshot: SnapshotAssertion,
):
    """Test generate command when no changes detected."""

    from metaxy.cli.migrations import generate
    from metaxy.config import MetaxyConfig

    # Create config with DuckDB store (not InMemory)
    config_file = tmp_path / "metaxy.toml"
    db_path = tmp_path / "test.duckdb"
    config_file.write_text(f"""
store = "test"

[stores.test]
type = "metaxy.metadata_store.duckdb.DuckDBMetadataStore"

[stores.test.config]
database = "{db_path}"
""")

    # Load config and create store with data
    config = MetaxyConfig.load(config_file)

    # Use test graph context
    with cli_graph.use():
        store = config.get_store("test")

        # Write data and call command in context manager
        with store:
            TestFeature = cli_graph.features_by_key[FeatureKey(["test_cli", "feature"])]
            data = pl.DataFrame(
                {
                    "sample_id": [1, 2],
                    "data_version": [{"default": "h1"}, {"default": "h2"}],
                }
            )
            store.write_metadata(TestFeature, data)
            store.record_feature_graph_snapshot()  # No arguments - records all features in graph

            set_config(config)
            generate(migrations_dir=Path("migrations"))

    # Check output
    captured = capsys.readouterr()
    assert captured.out.strip() == snapshot


def test_migrations_apply_dry_run(
    tmp_path: Path, cli_graph: FeatureGraph, capsys, snapshot: SnapshotAssertion
):
    """Test apply command in dry-run mode."""

    from metaxy.cli.migrations import apply
    from metaxy.config import MetaxyConfig
    from metaxy.migrations import DataVersionReconciliation, Migration

    # Create config with DuckDB (persistent store for CLI)
    db_path = tmp_path / "test.duckdb"
    config_file = tmp_path / "metaxy.toml"
    config_file.write_text(f"""
store = "test"

[stores.test]
type = "metaxy.metadata_store.duckdb.DuckDBMetadataStore"
config.database = "{db_path}"
""")

    config = MetaxyConfig.load(config_file)

    # Use test graph context
    with cli_graph.use():
        store = config.get_store("test")

        # Record a snapshot first so migration can load historical graph
        with store:
            snapshot_id, _ = store.record_feature_graph_snapshot()

            # Verify snapshot was recorded
            from metaxy.metadata_store.base import FEATURE_VERSIONS_KEY

            fv = collect_to_polars(
                store.read_metadata(FEATURE_VERSIONS_KEY, current_only=False)
            )
            assert len(fv) > 0, "No feature versions recorded"
            assert snapshot_id in fv["snapshot_id"].to_list(), (
                f"Snapshot {snapshot_id} not in table"
            )

        # Create test migration file
        migrations_dir = tmp_path / "migrations"
        migrations_dir.mkdir()

        migration = Migration(
            version=1,
            id="test_migration",
            parent_migration_id=None,
            from_snapshot_id=snapshot_id,
            to_snapshot_id=snapshot_id,
            description="Test",
            created_at=datetime(2025, 1, 1, 0, 0, 0),
            operations=[
                DataVersionReconciliation(
                    id="test_op",
                    feature_key=["test_cli", "feature"],
                    from_="abc123",
                    to="def456",
                    reason="Test migration",
                ).model_dump(by_alias=True)
            ],
        )
        migration.to_yaml(str(migrations_dir / "test_migration.yaml"))

        # Set context and call apply in context manager
        with store:
            set_config(config)
            apply(
                revision="test_migration",  # Migration ID instead of file
                migrations_dir=migrations_dir,
                dry_run=True,
                force=False,
            )

    # Check output with snapshot
    captured = capsys.readouterr()
    assert captured.out.strip() == snapshot


def test_migrations_status_empty(
    cli_config_file: Path,
    cli_graph: FeatureGraph,
    capsys,
    snapshot: SnapshotAssertion,
):
    """Test status command with no migrations."""
    from metaxy.cli.migrations import status
    from metaxy.config import MetaxyConfig

    config = MetaxyConfig.load(cli_config_file)

    # Use test graph context
    with cli_graph.use():
        store = config.get_store("test")

        # Call command in context manager
        with store:
            set_config(config)
            status()

    # Check output with snapshot
    captured = capsys.readouterr()
    assert captured.out.strip() == snapshot
