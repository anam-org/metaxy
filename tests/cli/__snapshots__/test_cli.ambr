# serializer version: 1
# name: test_migrations_apply_dry_run
  '''
  === DRY RUN MODE ===
  
  Applying 1 migration(s) to reach 'test_migration':
  
  âŠ˜ test_migration (already applied)
  
  Summary:
    Applied: 0
    Skipped: 1
    Failed: 0
  '''
# ---
# name: test_migrations_generate_no_changes
  '''
  From: latest snapshot 19ba17b377d925e17ba09aac663e8df5c37cd14509086bb061fd5a06e7945e01...
  To: current active graph (snapshot 19ba17b377d925e17ba09aac663e8df5c37cd14509086bb061fd5a06e7945e01... already recorded)
  No feature changes detected. All features up to date!
  '''
# ---
# name: test_migrations_status_empty
  'No migrations found.'
# ---
